packages/playground/data-liberation/src/xml-api/WP_XML_Processor.php

```diff
@@ -1865,36 +1880,116 @@ private function skip_whitespace() {
 		$this->bytes_already_parsed += strspn( $this->xml, " \t\f\r\n", $this->bytes_already_parsed );
 	}
 
-	// Describes the first character of the attribute name:
-	// NameStartChar ::= ":" | [A-Z] | "_" | [a-z] | [#xC0-#xD6] | [#xD8-#xF6] | [#xF8-#x2FF] | [#x370-#x37D] | [#x37F-#x1FFF] | [#x200C-#x200D] | [#x2070-#x218F] | [#x2C00-#x2FEF] | [#x3001-#xD7FF] | [#xF900-#xFDCF] | [#xFDF0-#xFFFD] | [#x10000-#xEFFFF]
-	// See https://www.w3.org/TR/xml/#NT-Name
-	const NAME_START_CHAR_PATTERN = ':a-z_A-Z\x{C0}-\x{D6}\x{D8}-\x{F6}\x{F8}-\x{2FF}\x{370}-\x{37D}\x{37F}-\x{1FFF}\x{200C}-\x{200D}\x{2070}-\x{218F}\x{2C00}-\x{2FEF}\x{3001}-\x{D7FF}\x{F900}-\x{FDCF}\x{FDF0}-\x{FFFD}\x{10000}-\x{EFFFF}';
-	const NAME_CHAR_PATTERN       = '\-\.0-9\x{B7}\x{0300}-\x{036F}\x{203F}-\x{2040}:a-z_A-Z\x{C0}-\x{D6}\x{D8}-\x{F6}\x{F8}-\x{2FF}\x{370}-\x{37D}\x{37F}-\x{1FFF}\x{200C}-\x{200D}\x{2070}-\x{218F}\x{2C00}-\x{2FEF}\x{3001}-\x{D7FF}\x{F900}-\x{FDCF}\x{FDF0}-\x{FFFD}\x{10000}-\x{EFFFF}';
+	/**
+	 * Parses a Name token starting at $offset
+	 * 
+	 * Name ::= NameStartChar (NameChar)*
+	 * 
+	 * @param int $offset
+	 * @return int
+	 */
 	private function parse_name( $offset ) {
-		if ( 1 !== preg_match(
-			'~[' . self::NAME_START_CHAR_PATTERN . ']~Ssu',
-			$this->xml[ $offset ],
-			$matches
-		) ) {
-			return 0;
-		}
+		$name_byte_length = 0;
+		while(true) {
```

Comment:
@adamziel @dmsnell @sirreal
As I'm looking into a similar problem right now, I thought I'd dump here an idea that I used in my case.

In the loop, we could 1) check for a sequence of ASCII < 128 characters, 2) check if the next character can be multibyte, and 3) only then call `utf8_codepoint_at`. In my scenario, it gives some performance gains. If we expect most attribute names to be ASCII < 128, then this could bring significant performance improvements.

(This is a simplification as I didn't consider the `$test_as_first_character` handling.)

```php
while (true) {
	// First, let's try to parse an ASCII sequence.
	$name_byte_length += strspn(
		$this->xml,
		'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ:_',
		$offset + $name_byte_length
	);

	// Check if the following byte can be part of a multibyte character.
	$byte = $this->xml[ $offset + $name_byte_length ] ?? null;
	if ( null === $byte || ord( $byte ) < 128 ) {
		break;
	}

	// Check the \x{0080}-\x{ffff} Unicode character range.
	$codepoint = utf8_codepoint_at( $this->xml, $offset + $name_byte_length, $bytes_parsed );
	if(
		null === $codepoint ||
		!$this->is_valid_name_codepoint( $codepoint, $name_byte_length === 0 )
	) {
		break;
	}
	$name_byte_length += $bytes_parsed;
}
```